### PR TITLE
Refactoring test

### LINKED ISSUE #24: Refactoring test
Refactor and unify feature test and step definitions.

### PR DESCRIPTION
**Pull request by j0sv**
Fixes #24

### Changes proposed by this PR

- Refactor and unify feature test and step definitions.

### CODE REVIEW

File: features/step_definitions/basic_step.rb [hunk 1]
```diff
@@ -1,16 +1,51 @@
...
+end
+
+#Background steps
+Given(/^the user "([^"]*)" exist$/) do |name|
+  FactoryGirl.create(:user, name: name)
+end
+
+Given(/^the following user exist$/) do |table|
+  table.hashes.each do | user |
+    User.create(user)
```
Comment (tochman) [inline features/step_definitions/basic_step.rb]: Again, Please use Factory Girl.

**Review by Philippoes — APPROVED** [review]
I think this is fine

File: features/compose_message.feature [hunk 2]
```diff
@@ -4,28 +4,25 @@ Feature: Compose messages
...
-    Given I am on landing page
-    And I click on the "Inbox" button
-    And I click on the "Compose" button
-    And I select "lisa" from the recipient list
-    And I enter "good" in "Body"
-    And I click on the "Send Message" button
+    Given I visit "landing" page
+    And I click "Inbox" button
+    And I click "Compose" button
+    And I select "lisa" from the "Recipient"
```
Comment (tochman) [inline features/compose_message.feature:15]: I would rephrase this to `I select "lisa" from "Recipient"`. Also, please use capital letters in names `lisa` -> `Lisa`

**Review by satorinic — APPROVED** [review]

File: features/step_definitions/basic_step.rb [hunk 3]
```diff
@@ -1,16 +1,45 @@
...
-  save_and_open_page
-end
-
+#Action steps
 When(/^I visit "([^"]*)" page$/) do |page_name|
   visit get_path page_name
 end
 
+#Form action steps
+When(/^I click "([^"]*)" (?:link|button)$/) do |link_or_button_name|
```
Comment (tochman) [inline features/step_definitions/basic_step.rb:7]: +1

File: features/step_definitions/basic_step.rb [hunk 4]
```diff
@@ -1,16 +1,45 @@
...
+  click_link_or_button link_or_button_name
+end
+
+When(/^I fill in "([^"]*)" with "([^"]*)"$/) do |field, value|
+  fill_in field, with: value
+end
+
+Given(/^I select "([^"]*)" from the "([^"]*)"$/) do |option, select_field|
```
Comment (tochman) [inline features/step_definitions/basic_step.rb:15]: I would remove `the`

File: features/step_definitions/basic_step.rb [hunk 5]
```diff
@@ -1,16 +1,45 @@
...
+  select option, from: select_field
+end
+
+#Assertions steps
+Then(/^I should see "([^"]*)"$/) do |content|
+  expect(page).to have_content content
+end
+
 Then(/^I should be on "([^"]*)" page$/) do |page_name|
   expect(page).to have_current_path get_path page_name
```
Comment (tochman) [inline features/step_definitions/basic_step.rb:25]: A bit out of scope for this PR but I want to comment on this. This is not readable. I would actually split this into two lines. 
```ruby 
path = get_path(page_name)
expect(page).to have_current_path path
```
Since you are calling an method to get the path I would argue that this is more readable.

File: features/step_definitions/basic_step.rb [hunk 6]
```diff
@@ -1,16 +1,45 @@
...
 end
 
+Then(/^show me the page$/) do
+  save_and_open_page
+end
```
Comment (tochman) [inline features/step_definitions/basic_step.rb:30]: This should not be grouped in assertion steps but rather in a separate group. Perhaps `#debugging steps`?

File: features/user_login.feature [hunk 7]
```diff
@@ -11,12 +11,12 @@ Scenario: User Login
...
+  Then I should see "Signed in successfully."
 
 Scenario: User Login with invalid password
   When I visit "landing" page
   And I click "Login" button
   And I fill in "Email" with "bob@bob.com"
   And I fill in "Password" with "bob1234S"
   And I click "Log in" button
-  Then I should see the message "Invalid Email or password."
+  Then I should see "Invalid Email or password."
```
Comment (tochman) [inline features/user_login.feature:22]: I actually preferred `Then I should see the message "Signed out successfully."`. Or perhaps refactor it to `Then I should see a flash message "Signed out successfully."` and make sure it is actually a css class (in your step definition)?
